File: golang/vertical_order_traversal_of_a_binary_tree.go
package golang

import (
	"golang/model"
	"sort"
)

type pair struct {
    Node *model.TreeNode
    Pos int
}

func max(a int, b int) int {
    if a > b {
        return a
    }
    return b
}

func verticalTraversal(root *model.TreeNode) [][]int {
    m := make(map[int][]int)
    var ans [][]int
    var stack []pair
    minv, maxv := 0, 0
    stack = append(stack, pair{
        Node: root,
        Pos: 0,
    })
    
    for len(stack) > 0 {
        mtemp := make(map[int][]int)
        var atemp []pair
        
        for _, item := range stack {
            minv = min(minv, item.Pos)
            maxv = max(maxv, item.Pos)
            
            if _, ok := mtemp[item.Pos]; ok {
                mtemp[item.Pos] = append(mtemp[item.Pos], item.Node.Val)
            } else {
                mtemp[item.Pos] = []int{item.Node.Val}
            }
            
            if item.Node.Left != nil {
                atemp = append(atemp, pair{
                    Node: item.Node.Left,
                    Pos: item.Pos - 1,
                })
            }
            
            if item.Node.Right != nil {
                atemp = append(atemp, pair{
                    Node: item.Node.Right,
                    Pos: item.Pos + 1,
                })
            }
        }
        
        for key, val := range mtemp {
            sort.Ints(val)
            if _, ok := m[key]; ok {
                m[key] = append(m[key], val...)
            } else {
                m[key] = val
            }
        }
        
        stack = atemp
    }
    
    for i := minv; i <= maxv; i++ {
        ans = append(ans, m[i])
    }
    
    return ans
}

File: python3/vertical_order_traversal_of_a_binary_tree.py
from .model import TreeNode

from collections import defaultdict

"""
Space   : O(n)
Time    : O(n log n)
"""


class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        d = defaultdict(list)
        stack = [(root, 0)]
        minv, maxv = 0, 0

        while stack:
            stemp = []
            dtemp = defaultdict(list)
            for node, pos in stack:
                minv = min(minv, pos)
                maxv = max(maxv, pos)

                dtemp[pos].append(node.val)

                if node.left:
                    stemp.append((node.left, pos - 1))
                if node.right:
                    stemp.append((node.right, pos + 1))

            for k, v in dtemp.items():
                d[k].extend(sorted(v))

            stack = stemp

        ans = []
        for i in range(minv, maxv+1):
            ans.append(d[i])

        return ans
